avro-fastserde/src/main/java/com/linkedin/avro/fastserde/FastDeserializerGenerator.java

```diff
@@ -492,21 +494,41 @@ private void processUnion(JVar unionSchemaVar, final String name, final Schema u
       final Schema readerUnionSchema, JBlock body, FieldAction action,
       BiConsumer<JBlock, JExpression> putValueIntoParent, Supplier<JExpression> reuseSupplier) {
     JVar unionIndex = body.decl(codeModel.INT, getUniqueName("unionIndex"), JExpr.direct(DECODER + ".readIndex()"));
-    JConditional ifBlock = null;
+    JSwitch switchBlock = body._switch(unionIndex);
     for (int i = 0; i < unionSchema.getTypes().size(); i++) {
       Schema optionSchema = unionSchema.getTypes().get(i);
       Schema readerOptionSchema = null;
+      int readerOptionUnionBranchIndex = -1;
       FieldAction unionAction;
-      JExpression condition = unionIndex.eq(JExpr.lit(i));
+      JBlock caseBody = switchBlock._case(JExpr.lit(i)).body();
 
       if (Schema.Type.NULL.equals(optionSchema.getType())) {
-        ifBlock = ifBlock != null ? ifBlock._elseif(condition) : body._if(condition);
-        ifBlock._then().directStatement(DECODER + ".readNull();");
+        caseBody.directStatement(DECODER + ".readNull();");
```

Comment:
We cannot assume that the reader and writer will contain the same number of branches. I don't think we assume that the reader must contain null... here we are enumerating on the writer schema's union branches and providing a case (previously an if) for each branch. This code is transformed from if/else to switch/case but is logically equivalent, it will just invoke `decoder.readNull()` (which is a no-op, BTW).